File: scripts/create_desktop_icons.py
#!/usr/bin/env python3
"""Create desktop installer icons from the QuoteBook PNG icon."""

from __future__ import annotations

import struct
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "frontend" / "assets" / "app-icon-512.png"
BUILD = ROOT / "desktop" / "build"


def create_icns(png_bytes: bytes) -> bytes:
    icon_type = b"ic09"
    icon = icon_type + struct.pack(">I", len(png_bytes) + 8) + png_bytes
    return b"icns" + struct.pack(">I", len(icon) + 8) + icon


def create_ico(png_bytes: bytes) -> bytes:
    header = struct.pack("<HHH", 0, 1, 1)
    directory = struct.pack("<BBBBHHII", 0, 0, 0, 0, 1, 32, len(png_bytes), 22)
    return header + directory + png_bytes


def main() -> None:
    BUILD.mkdir(parents=True, exist_ok=True)
    png_bytes = SOURCE.read_bytes()
    (BUILD / "icon.icns").write_bytes(create_icns(png_bytes))
    (BUILD / "icon.ico").write_bytes(create_ico(png_bytes))


if __name__ == "__main__":
    main()

File: scripts/create_desktop_icons.js
const fs = require("node:fs");
const path = require("node:path");

const root = path.resolve(__dirname, "..");
const source = path.join(root, "frontend", "assets", "app-icon-512.png");
const build = path.join(root, "desktop", "build");

function uint32BE(value) {
  const buffer = Buffer.alloc(4);
  buffer.writeUInt32BE(value);
  return buffer;
}

function uint16LE(value) {
  const buffer = Buffer.alloc(2);
  buffer.writeUInt16LE(value);
  return buffer;
}

function uint32LE(value) {
  const buffer = Buffer.alloc(4);
  buffer.writeUInt32LE(value);
  return buffer;
}

function createIcns(png) {
  const icon = Buffer.concat([Buffer.from("ic09"), uint32BE(png.length + 8), png]);
  return Buffer.concat([Buffer.from("icns"), uint32BE(icon.length + 8), icon]);
}

function createIco(png) {
  const header = Buffer.concat([uint16LE(0), uint16LE(1), uint16LE(1)]);
  const directory = Buffer.concat([
    Buffer.from([0, 0, 0, 0]),
    uint16LE(1),
    uint16LE(32),
    uint32LE(png.length),
    uint32LE(22),
  ]);
  return Buffer.concat([header, directory, png]);
}

fs.mkdirSync(build, { recursive: true });
const png = fs.readFileSync(source);
fs.writeFileSync(path.join(build, "icon.icns"), createIcns(png));
fs.writeFileSync(path.join(build, "icon.ico"), createIco(png));
